include/LightGBM/category_encoding_provider.hpp

```diff
@@ -0,0 +1,771 @@
+          char_pointer += line_len;
+          char_pointer = Common::SkipNewLine(char_pointer);
+          if (!Common::StartsWith(line.c_str(), "prior=")) {
+            Log::Fatal("CatConverter model format error");
+          }
+          double prior = 0.0f;
+          Common::Atof(Common::Split(line.c_str(), "=")[1].c_str(), &prior);
+          if (type == std::string("target_encoder")) {
+            ret = new TargetEncoder(prior);
+          } else {
+            ret = new TargetEncoderLabelMean();
+          }
+          ret->SetPrior(prior, prior_weight);
+        } else if (type == std::string("count_encoder")) {
+          ret = new CountEncoder();
+        } else {
+          Log::Fatal("Unknown CatConverter type %s", type.c_str());
+        }
+
+        line_len = Common::GetLine(char_pointer);
+        line = std::string(char_pointer, line_len);
+        char_pointer += line_len;
+        char_pointer = Common::SkipNewLine(char_pointer);
+        if (!Common::StartsWith(line.c_str(), "categorical_feature_index_to_encoded_feature_index=")) {
+          Log::Fatal("CatConverter model format error");
+        }
+        std::vector<std::string> feature_index_pair = Common::Split(Common::Split(line.c_str(), "=")[1].c_str(), " ");
+        ret->cat_fid_to_convert_fid_.clear();
+        for (auto& pair_string : feature_index_pair) {
+          std::vector<std::string> cat_fid_and_convert_fid_string = Common::Split(pair_string.c_str(), ":");
+          int cat_fid = 0;
+          Common::Atoi(cat_fid_and_convert_fid_string[0].c_str(), &cat_fid);
+          int convert_fid = 0;
+          Common::Atoi(cat_fid_and_convert_fid_string[1].c_str(), &convert_fid);
+          ret->cat_fid_to_convert_fid_[cat_fid] = convert_fid;
+        }
+        *used_len = static_cast<size_t>(char_pointer - char_pointer_start);
+      } else {
+        Log::Fatal("CatConverter model format error");
+      }
+      return ret;
+    }
+  };
+
+  class TargetEncoder: public CatConverter {
+   public:
+    explicit TargetEncoder(const double prior): prior_(prior) {}
+
+    inline double CalcValue(const double sum_label, const double sum_count,
+      const double /*all_fold_sum_count*/) const override {
+      return (sum_label + prior_ * prior_weight_) / (sum_count + prior_weight_);
+    }
+
+    inline double CalcValue(const double sum_label, const double sum_count,
+      const double /*all_fold_sum_count*/, const double /*prior*/) const override {
+      return (sum_label + prior_ * prior_weight_) / (sum_count + prior_weight_);
+    }
+
+    void SetPrior(const double /*prior*/, const double prior_weight) override {
+      prior_weight_ = prior_weight;
+    }
+
+    std::string FeatureName() const override {
+      std::stringstream str_stream;
+      Common::C_stringstream(str_stream);
+      str_stream << "target_encoding_" << prior_;
+      return str_stream.str();
+    }
+
+    json11::Json DumpToJSONObject() const override {
```

Comment:
Looks like it is another serialization logic for sub/base class, could we have pattern like below: 

DumpToJSONObject(json11::Json context)
{
     base.DumpToJSONObject(context);

     // Dump to json logic for sub class
}